### PR TITLE
feat: portal slippage optimization approval swaps

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

https://github.com/shapeshift/web/pull/7654 added auto slippage support for Portals trades that could be validated, i.e. those that did not require approvals.

This PR closes the loop by supporting Portals trades that _do_ require approval by using the `[v2/portal/estimate](https://api.portals.fi/docs#/Portal/PortalController_estimatePortalV2)` endpoint.

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/7650

## Risk

Medium - adds an extra request to get the optimal slippage from the Portals API.

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

Portals trades.

## Testing

Get a quote for a Portals supported trade pair where the sell asset (an ERC20) does not have enough allowance approval to perform the trade.

Screenshot 2024-09-02 at 7 15 18 PM

Ensure that we get a quote with an automated slippage value (between 0 and 2.5%).
Ensure that we can complete the trade.

### Engineering

In the cases where the standard quote fails due to missing allowance approval we'll see the following console log:

Screenshot 2024-09-02 at 4 43 00 PM

In the network tab we can expect to see 4 requests:

Screenshot 2024-09-02 at 4 50 24 PM

The first will fail, the second gets the slippage value, the third gets the gas, and the 4th gets the quote to execute.

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

☝️

## Screenshots (if applicable)

N/A

### CODE REVIEW

File: packages/swapper/src/constants.ts [hunk 1]
```diff
@@ -91,7 +91,6 @@ export const swappers: Record<
 // Slippage defaults. Don't export these to ensure the getDefaultSlippageDecimalPercentageForSwapper helper function is used.
 const DEFAULT_SLIPPAGE_DECIMAL_PERCENTAGE = '0.002' // .2%
 const DEFAULT_COWSWAP_SLIPPAGE_DECIMAL_PERCENTAGE = '0.005' // .5%
-const DEFAULT_PORTALS_SLIPPAGE_DECIMAL_PERCENTAGE = '0.01' // 1%
```
Comment (gomesalexandre) [inline packages/swapper/src/constants.ts:94]: 👋🏽

File: packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts [hunk 2]
```diff
@@ -120,10 +119,24 @@ export async function getPortalsTradeQuote(
...
-      // 1. a quote with validation enabled, but using a well-funded address to get a rough gasLimit estimate
-      // 2. another quote with validation disabled, to get an actual quote
+      // If validation fails, fire 3 more quotes:
+      // 1. a quote estimate (does not require approval) to get the optimal slippage tolerance
+      // 2. a quote with validation enabled, but using a well-funded address to get a rough gasLimit estimate
+      // 3. another quote with validation disabled, to get an actual quote (using the user slippage, or the optimal from the estimate)
       console.info('failed to get Portals quote with validation enabled', e)
+
+      // Use the quote estimate endpoint to get the optimal slippage tolerance
+      const quoteEstimateResponse = await fetchPortalsTradeEstimate({
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:129]: Confirmed she's the goer!

image

File: packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts [hunk 3]
```diff
@@ -132,6 +145,7 @@ export async function getPortalsTradeQuote(
       const dummyInputToken = `${portalsNetwork}:${dummySellAssetAddress}`
       const dummyOutputToken = `${portalsNetwork}:${dummyBuyAssetAddress}`
 
+      // Use a dummy request to the portal endpoint to get a rough gasLimit estimate
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:148]: 💜

File: packages/swapper/src/swappers/PortalsSwapper/utils/fetchPortalsTradeOrder.ts [hunk 4]
```diff
@@ -50,6 +52,25 @@ type PortalsTradeOrderResponse = {
...
+  context: {
+    slippageTolerancePercentage: number
+    inputAmount: string
+    inputAmountUsd: number
+    inputToken: string
+    outputToken: string
+    outputAmount: string
+    outputAmountUsd: number
+    minOutputAmountUsd: number
+    sender: string
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/PortalsSwapper/utils/fetchPortalsTradeOrder.ts:70]: This should be optional according to the swagger spec

image

**Review by gomesalexandre — COMMENTED** [review]
Tested locally, confirmed this does what it says on the box!

Added one comment re: default slippage safety, happy to get this one in when tackled 🙏🏽 

- Patched `quoteEstimateResponse` to `undefined` to simulate failing `/v2/portal/estimate` calls on the approval needed path and got an upstream error on the `/v2/portal` call

image

- Without monkey patch applied, confirmed we are still happy with an approval required

image
image

- And with approval already granted too

image

File: packages/swapper/src/constants.ts (hunk 1 above)
Comment (0xApotheosis) [reply packages/swapper/src/constants.ts:94]: My assumption was that if we can't get a successful `/v2/estimate` then we probably can't proceed anyway, but I can add it in as another fallback.

File: packages/swapper/src/swappers/PortalsSwapper/utils/fetchPortalsTradeOrder.ts (hunk 4 above)
Comment (0xApotheosis) [reply packages/swapper/src/swappers/PortalsSwapper/utils/fetchPortalsTradeOrder.ts]: https://github.com/shapeshift/web/pull/7672/commits/102812593a73ebd529202b70c88a65b76f9f0f5c 👌

File: packages/swapper/src/constants.ts (hunk 1 above)
Comment (0xApotheosis) [reply packages/swapper/src/constants.ts]: Addressed in https://github.com/shapeshift/web/pull/7672/commits/102812593a73ebd529202b70c88a65b76f9f0f5c 👌

**Review by gomesalexandre — APPROVED** [review]
Restested locally:

- Monkey patched `/v2/estimate` failures, with dummy quote needed

image
image
image

- No monkey-patch/no dummy quote needed

image

**Pull request merged**